src/OauthTokenFileStorage.php

```diff
@@ -0,0 +1,193 @@
+use Drupal\Core\Site\Settings;
+
+/**
+ * Storing and returning OAuth access token data.
+ */
+class OauthTokenFileStorage implements OauthTokenStorageInterface {
+
+  const OAUTH_TOKEN_FILE_PATH_SETTINGS_KEY = 'apigee_edge_oauth_token_file_path';
+
+  /**
+   * Ensures that token gets refreshed earlier than it expires.
+   *
+   * Number of seconds extracted from token's expiration date when
+   * hasExpired() calculates.
+   *
+   * @var int
+   */
+  protected $leeway = 30;
+
+  /**
+   * An internally cached token data store.
+   *
+   * @var array
+   */
+  private static $token_data;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAccessToken(): ?string {
+    $token_data = $this->getTokenData();
+    return $token_data['access_token'] ?? NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getTokenType(): ?string {
+    $token_data = $this->getTokenData();
+    return $token_data['token_type'] ?? NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRefreshToken(): ?string {
+    $token_data = $this->getTokenData();
+    return $token_data['refresh_token'] ?? NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getScope(): string {
+    $token_data = $this->getTokenData();
+    return $token_data['scope'] ?? NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpires(): int {
+    $token_data = $this->getTokenData();
+    return $token_data['expires'] ?? -1;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function hasExpired(): bool {
+    $expires = $this->getExpires();
+    if ($expires !== 0 && ($expires - $this->leeway) > time()) {
+      return FALSE;
+    }
+
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function markExpired(): void {
+    // Gets token data.
+    $token_data = $this->getTokenData();
+    // Expire in the past.
+    $token_data['expires_in'] = -1;
+    // Save the token data.
+    $this->saveToken($token_data);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function saveToken(array $data): void {
+    // Calculate the cache expiration.
+    $data['expires'] = isset($data['expires_in']) ? $data['expires_in'] + time() : ($data['expires'] ?? -1);
+    // Remove the expires_in data.
+    unset($data['expires_in']);
+
+    // Gets the file directory so we can make sure it exists.
+    $file_path = dirname(static::oauthTokenPath());
+    file_prepare_directory($file_path, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS);
+
+    // Write the obfuscated token data to a private file.
+    file_unmanaged_save_data(base64_encode(serialize($data)), static::oauthTokenPath(), FILE_EXISTS_REPLACE);
+
+    // Update the cached value.
+    static::$token_data = $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function removeToken(): void {
+    // Remove the data from the static cache.
+    static::$token_data = NULL;
+    // Remove the token data from storage.
+    file_unmanaged_delete(static::oauthTokenPath());
```

Comment:
Maybe this should not delete the file rather clear its content only. Just like `saveToken([])` does.

Reply:
Good idea